suggestions

tlambert03 opened the pull request:
if you merge the current #1444 here's what I was starting to do to convert the layerlist.
I really think we want to get away from having `LayerList` list to it's own event to call its own method, so this removes that `self.events.inserted.connect(self._add)` connection from the `__init__` method, and instead reimplements the two methods where items can actually be added to the list (`__setitem__` and `insert`) to make sure we do a little prep (`_prepare_to_add_layer`) before calling the `super()` method.

One test on the layerlist is still broken around selection logic in `move_selected`, I worked on it for a while but it's has the potential to touch too many things... so I'm leaving it out of #1444 and you can keep working on it here

tlambert03 committed:
suggestions

tlambert03 commented on napari/components/_tests/test_layers_list.py:
```diff
@@ -102,16 +102,16 @@ def test_reordering():
-    # Swap layers by name
-    layers['image_b', 'image_c'] = layers['image_c', 'image_b']
-    assert list(layers) == [layer_c, layer_a, layer_b]
+    # # Swap layers by name
```
just dropping this syntax

tlambert03 commented on napari/components/layerlist.py:
```diff
@@ -1,36 +1,50 @@
-        layer.events.name.connect(lambda e: self._update_name(e))
-        self.unselect_all(ignore=layer)
+    def _prepare_to_add_layer(self, layer: Layer):
+        if isinstance(layer, Layer):
```
this happens before the type enforcement... so we still have to check if it's a layer.

sofroniewn reviewed (approved):
LGTM!

I definitely want to drop `move_selected` - that was one of the nice things done in #1391 before we reverted it.

I will say I find it a bit unfortunate now that the reordering seems to emit inserted events unless we block them which it didn't used to

sofroniewn commented on napari/components/layerlist.py:
```diff
@@ -1,36 +1,50 @@
+    def __setitem__(self, key, value):
+        self._prepare_to_add_layer(value)
+        super().__setitem__(key, value)
+        # FIXME: the behavior of unselecting everything but the last-added
```
I wouldn't call the unselection behavior error prone - it worked fine until what has been a major change in how events are emitted. I think I still want the behavior that adding a new layer leads to only it being selected, but overwriting/changing an existing one doesn't need to change selection - it should though do the name checking which it wasn't before though, and now is

tlambert03 replied:
sorry, that was poor wording on my part.  I agree with you, the "behavior" is fine.   By "error" prone i meant that you need to change the list "just the right way" for it to work.  So perhaps i mean it's a bit brittle.  The intention here isn't to have a major change in the how the events are emitted... just to make sure that they aren't emitted selectively, like when you use `pop()` or `insert()` but not when you use `clear()` or slice style insertion ... We've gone through effort to make the layer list behavior like a "normal" list, so we have to expect/prepare for users using _any_ of the valid list methods (_or_ we have to make them specifically inaccessible)

sofroniewn commented on napari/components/layerlist.py:
```diff
@@ -98,7 +114,9 @@ def move_selected(self, index, insert):
             indices.insert(insert_idx, elem_idx)
         # reorder list
-        self[:] = self[tuple(indices)]
+        with self.events.blocker():
```
It looks like the events is `qt_layerlist.py` are still being triggered despite this block https://github.com/napari/napari/pull/1504/files#diff-6408ff7cd9e06a487f66df8d17583e6eR64-R66 not sure why though

sofroniewn replied:
The blocker does seem to stop event listeners added after the layerlist is created, say in the console, so not quite sure what's happening here

tlambert03 replied:
i'll take a look... can you point me to code I can use to reproduce seeing the `qt_layerlist` emit events despite the block?

sofroniewn replied:
Not at my keyboard right now, but I added print statements inside the listeners in that file (the two where layer widgets get removed / added respectively) then loaded a napari viewer with a bunch of layers, did a manual drag and drop and saw those print statements appear inside the blocker context - which to me was unexpected

sofroniewn replied:
print statements were https://github.com/napari/napari/blob/master/napari/_qt/qt_layerlist.py#L95 and here https://github.com/napari/napari/blob/master/napari/_qt/qt_layerlist.py#L79, but i also see it with print statements here https://github.com/napari/napari/blob/master/napari/_qt/qt_controls.py#L66 and https://github.com/napari/napari/blob/master/napari/_qt/qt_controls.py#L79.

If I do `viewer.layers.events.connect(lambda e: print(f"you called {e.type}"))` then this print statement is blocked by the blocker.

I wonder if the blocker blocks `viewer.layers.events.connect` but doesn't block individual ``viewer.layers.events.insertedconnect` calls. I will test now!!

sofroniewn replied:
Yes, this appears to be the case `with self.events.blocker():` will block

```python
viewer.layers.events.connect(lambda e: print(f"you called {e.type}"))
```

but not block

```python
viewer.layers.events.inserted.connect(lambda e: print(f"you called ONLY INSERTED"))
```

I think this is a bug and it should block that as well. I think with that fixed this PR would then work as expected.

Thoughts @tlambert03 - can you reproduce?

tlambert03 replied:
I can reproduce this.  let me figure this one out before we try to work around it

tlambert03 commented:
> I will say I find it a bit unfortunate now that the reordering seems to emit inserted events unless we block them which it didn't used to

:) i honestly see that as more of a "feature" than a bug.  The whole point here is that the previous model was "leaky": emitting events with some mutation events but not with others.  That's how stuff like ~https://github.com/napari/napari/issues/1467~ https://github.com/napari/napari/issues/1380 happens... where basically our tests were "working", but only because we didn't anticipate someone mutating the list using any methods other than the ones we happen to use internally.  To me, it's preferable (and normal) to first make sure it's very hard to touch the list "secretly" without an event, and only then block it every so often in special circumstances (like this one... where we only block it because we're about to emit a different event once the mutation is complete).

tlambert03 commented:
sorry... I meant to link https://github.com/napari/napari/issues/1380 above... not 1467

sofroniewn commented:
> :) i honestly see that as more of a "feature" than a bug. The whole point here is that the previous model was "leaky": emitting events with some mutation events but not with others. That's how stuff like napari#1467 napari#1380 happens... where basically our tests were "working", but only because we didn't anticipate someone mutating the list using any methods other than the ones we happen to use internally. To me, it's preferable (and normal) to first make sure it's very hard to touch the list "secretly" without an event, and only then block it every so often in special circumstances (like this one... where we only block it because we're about to emit a different event once the mutation is complete).

I'm still not sure. i super agree with "so we have to expect/prepare for users using any of the valid list methods" which you say above, but I think the reordering case still might be a little special, because it seems like its not really a normal list method.

Now with a reorder it looks like we remove every element and then add every element back in, which means we're creating and destroying every qt list widget every time there's any drag and drop. Even if two elements are swapped we create and destroy every one.

Internally now we use the blocker, but if someone did `viewer.layers[:] = [viewer.layers[i] for i in (0, 2, 4, 3, 1)]` without the blocker then things would break because we'd creating/ destroying all these qt objects, right?

You've added `move` events and I feel like if we must I'd rather we trigger a move event https://github.com/napari/napari/pull/1444/files#diff-e1f85a058e2b79cbcc60571ed3d2ae68R68-R71
which emits 
```python
    moved (index: int, new_index: int, value: T)
        emitted after ``value`` is moved from ``index`` to ``new_index``
```
How about we allow `index` and `new_index` to be slices? or tuples? so that I can do a reorder much more simply and just listen the to `moved` event. We could then drop the `reodered` event entirely. Even a `reverse` could emit a `moved` event.

How does that sound?

tlambert03 commented:
this is all good stuff that gets right to the heart of the purpose for refactor so it's worth diving into this.

>  but I think the reordering case still might be a little special, because it seems like it's not really a normal list method.

reordering is definitely the most complicated case when it comes to events.  and there are multiple places or levels at which we could choose to address it.

First, it _would_ be possible to make the `viewer.layers[:] = [viewer.layers[i] for i in (0, 2, 4, 3, 1)]` syntax emit moved events, but I don't think we should.  It would require making the [`__setitem__` method](https://github.com/napari/napari/blob/214256bfb05193abc9cd1b77f1026460a7875061/napari/utils/events/containers/_list.py#L133) very heavy (and it's already starting to get bloated), basically saying: "anytime the `key` of the `__setitem__` operation is a slice object, lets first store the set of objects that currently occupy that slice (just in case we see them again), then find the intersection between those objects and the incoming objects objects and emit moved events for those, and then find the difference between the outgoing objects and the incoming objects and emit removed events for those, and emit inserted events for everything else that's incoming."

We don't want that kind of overhead everytime we use a slice operation (I don't think?), so we offer higher level move methods with which the user can indicate "don't worry, they're coming right back... you don't need to compare every object" and we can pause events internally and emit a moved event when the operation is done.  See [`EventedList.move`](https://github.com/napari/napari/blob/214256bfb05193abc9cd1b77f1026460a7875061/napari/utils/events/containers/_list.py#L204) and [`NestableEventedList.move_multiple`](https://github.com/napari/napari/blob/214256bfb05193abc9cd1b77f1026460a7875061/napari/utils/events/containers/_nested_list.py#L258).  This all gets much more complicated with trees (where you could be moving from multiple arbitrary levels to any other level), so using a dedicated method like `move_multiple` is really the only option (see [here](https://github.com/napari/napari/blob/4c22fe82b03586f00f6c6db9c43d43f1529b7783/napari/_qt/tree/_tree_model.py#L90) for where this gets called by the `AbstractItemModel` during a drag and drop in the layergroup branch).  Note, I don't have a `move_multiple` method on the simple list yet, because I was focusing on the more complicated layergroup model, but I can add one there too.

> Now with a reorder it looks like we remove every element and then add every element back in, which means we're creating and destroying every qt list widget every time there's any drag and drop.  Internally now we use the blocker, but if someone did viewer.layers[:] = [viewer.layers[i] for i in (0, 2, 4, 3, 1)] without the blocker then things would break because we'd creating/ destroying all these qt objects, right?

 Yes, if someone used the slice syntax instead of the move methods it would destroy then create.  But I I don't see why that should break anything unless we've done a bad job elsewhere in the code.  Those qt controls are just views onto the model, and they are _cheap_ views at that, taking only about 2 ms to build.  And in MV, there should be no problem having multiple views onto a model, or removing one and making a new one.  So the _worst_ thing that should happen is that the user incurs an extra 2ms per layer being moved if they use slice syntax rather than `layer_list.move()`.  If it breaks something, then we have issues in our Qt layers, not the model itself.

> I feel like if we must I'd rather we trigger a move event

the move event is triggered by the [`EventedList.move`](https://github.com/napari/napari/blob/214256bfb05193abc9cd1b77f1026460a7875061/napari/utils/events/containers/_list.py#L204) method...  I was trying to keep the changes to the layergroup PR minimal at first, but it might be better to use that method in the `move_selected` method rather than sticking with the `viewer.layers[:] = [viewer.layers[i] for i in (0, 2, 4, 3, 1)]` syntax.  That syntax is really mostly coming from the old list model, where it "worked" (using the old tuple indexing syntax) ... but it only worked because a whole bunch of other operations were essentially broken and untested (napari/napari#1380).  If you need a `move_multiple` method on the basic evented list lemme know!

tlambert03 commented:
ok, I just added a `move_multiple` method to the basic `EventedList` as well.  so now:
```python
viewer.layers[:] = [viewer.layers[i] for i in (0, 2, 4, 3, 1)]
# is the same as 
viewer.layers.move_multiple((0, 2, 4, 3, 1), 0)
```
That will emit three events (you don't have to listen to them all):
```python
moving(index=(0, 2, 4, 3, 1), new_index=0)
moved(index=(0, 2, 4, 3, 1), new_index=0, value=list_of_layers)
reordered(value=list_of_layers)
# where list_of_layers is the actual layer_list after the move operation
```

sofroniewn commented:
> this is all good stuff that gets right to the heart of the purpose for refactor so it's worth diving into this.

agreed - ok - I think i'm getting it now and have really come around to your proposal.

Probably one of the most compelling points is that the `viewer.layers[:] = [viewer.layers[i] for i in (0, 2, 4, 3, 1)] ` syntax doesn't only need to be for reorders, that original list could have 10 elements in it, and that tuple could have repeats and we should be robust enough to work in all those cases.

> Yes, if someone used the slice syntax instead of the move methods it would destroy then create. But I I don't see why that should break anything unless we've done a bad job elsewhere in the code. Those qt controls are just views onto the model, and they are cheap views at that, taking only about 2 ms to build. And in MV, there should be no problem having multiple views onto a model, or removing one and making a new one. So the worst thing that should happen is that the user incurs an extra 2ms per layer being moved if they use slice syntax rather than layer_list.move(). If it breaks something, then we have issues in our Qt layers, not the model itself.

Yup I agree here, this should work regardless and we should make sure it's tested.

> ok, I just added a move_multiple method to the basic EventedList as well.

I really like this and think we should use it internally. I'll get https://github.com/napari/napari/pull/1504 updated and working with this new method and make sure all the other stuff works too. Then I think that PR will be done, and we can punt all this to @jni for a review (no need to go though this whole discussion @jni, but I think it was quite helpful for me!)
